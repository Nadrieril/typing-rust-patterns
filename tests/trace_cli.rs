use assert_cmd::Command;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

#[test]
fn trace_cli() -> anyhow::Result<()> {
    let test_cases = &[
        "quit",
        "help",
        "options",
        "set",
        "set mistyped",
        "set mistyped val",
        "set ref_binding_on_inherited true",
        "set allow_ref_pat_on_ref_mut false",
        "set downgrade_shared_inside_shared true",
        "set ergo2024\noptions",
        "&x: &mut T",
        "set allow_ref_pat_on_ref_mut false\n&x: &mut T",
        "rules",
    ];
    for &test_case in test_cases {
        let success = Command::cargo_bin("typing-rust-patterns")?
            .write_stdin(test_case)
            .assert()
            .success();
        let out = String::from_utf8(success.get_output().stdout.clone())?;

        // Identify each snapshot file by the hash of the test case.
        let req_hash = {
            let mut hasher = DefaultHasher::new();
            test_case.hash(&mut hasher);
            hasher.finish().to_string()
        };
        insta::with_settings!({
            snapshot_suffix => req_hash,
            info => &test_case,
            omit_expression => true,
            prepend_module_to_snapshot => true,
        }, {
            insta::assert_snapshot!(out);
        });
    }
    Ok(())
}
