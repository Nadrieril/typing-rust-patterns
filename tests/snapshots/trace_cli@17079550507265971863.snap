---
source: tests/trace_cli.rs
info: set inherited_ref_on_ref EatInner
---
The two rulesets are described by the following sets of rules, with differences highlighted.

 real, m ⊢ p0: T0,  real, m ⊢ p1: T1                                              | real, m ⊢ p0: T0,  real, m ⊢ p1: T1
 --------------------------------------- "Constructor"                            | --------------------------------------- "Constructor"
 _, m ⊢ [p0, p1]: [T0, T1]                                                        | _, m ⊢ [p0, p1]: [T0, T1]

 inh, ro ⊢ p0: &T0,  inh, ro ⊢ p1: &T1                                            | inh, ro ⊢ p0: &T0,  inh, ro ⊢ p1: &T1
 ----------------------------------------- "ConstructorRef"                       | ----------------------------------------- "ConstructorRef"
 r, m ⊢ [p0, p1]: &[T0, T1]                                                       | r, m ⊢ [p0, p1]: &[T0, T1]

 inh, m ⊢ p0: &mut T0,  inh, m ⊢ p1: &mut T1                                      | inh, m ⊢ p0: &mut T0,  inh, m ⊢ p1: &mut T1
 ----------------------------------------------- "ConstructorRef"                 | ----------------------------------------------- "ConstructorRef"
 r, m ⊢ [p0, p1]: &mut [T0, T1]                                                   | r, m ⊢ [p0, p1]: &mut [T0, T1]

 inh, ro ⊢ [p0, p1]: &T                                                           | inh, ro ⊢ [p0, p1]: &T
 ------------------------ "ConstructorMultiRef"                                   | ------------------------ "ConstructorMultiRef"
 r, m ⊢ [p0, p1]: &&T                                                             | r, m ⊢ [p0, p1]: &&T

 inh, ro ⊢ [p0, p1]: &T                                                           | inh, ro ⊢ [p0, p1]: &T
 -------------------------- "ConstructorMultiRef"                                 | -------------------------- "ConstructorMultiRef"
 r, m ⊢ [p0, p1]: &&mut T                                                         | r, m ⊢ [p0, p1]: &&mut T

 inh, ro ⊢ [p0, p1]: &T                                                           | inh, ro ⊢ [p0, p1]: &T
 -------------------------- "ConstructorMultiRef"                                 | -------------------------- "ConstructorMultiRef"
 r, m ⊢ [p0, p1]: &mut &T                                                         | r, m ⊢ [p0, p1]: &mut &T

 inh, m ⊢ [p0, p1]: &mut T                                                        | inh, m ⊢ [p0, p1]: &mut T
 ------------------------------ "ConstructorMultiRef"                             | ------------------------------ "ConstructorMultiRef"
 r, m ⊢ [p0, p1]: &mut &mut T                                                     | r, m ⊢ [p0, p1]: &mut &mut T

 real, ro ⊢ p: T                                                                  | real, ro ⊢ p: T
 ------------------ "Deref"                                                       | ------------------ "Deref(EatOuter)"
 real, m ⊢ &p: &T                                                                 | real, m ⊢ &p: &T

 _, ro ⊢ p: T                                                                     | _, ro ⊢ p: T
 --------------------------------------- "Deref"                                  | --------------------------------------- "Deref(EatOuter)"
 inh, m ⊢ &p: &T, T is not a reference                                            | inh, m ⊢ &p: &T, T is not a reference

 real, ro ⊢ p: &T                                                                 | inh, ro ⊢ p: &T
 ------------------ "Deref"                                                       | ------------------ "Deref(EatInner)"
 inh, m ⊢ &p: &&T                                                                 | inh, m ⊢ &p: &&T

 real, ro ⊢ p: &mut T                                                             | inh, ro ⊢ p: &T
 ---------------------- "Deref"                                                   | ---------------------- "DerefMutWithShared(EatInner)"
 inh, m ⊢ &p: &&mut T                                                             | inh, m ⊢ &p: &&mut T

 real, m ⊢ p: T                                                                   | real, m ⊢ p: T
 -------------------------- "Deref"                                               | -------------------------- "Deref(EatOuter)"
 real, m ⊢ &mut p: &mut T                                                         | real, m ⊢ &mut p: &mut T

 _, m ⊢ p: T                                                                      | _, m ⊢ p: T
 ----------------------------------------------- "Deref"                          | ----------------------------------------------- "Deref(EatOuter)"
 inh, m ⊢ &mut p: &mut T, T is not a reference                                    | inh, m ⊢ &mut p: &mut T, T is not a reference

 real, m ⊢ p: &T                                                                  |  
 -------------------------- "Deref"                                               |  
 inh, m ⊢ &mut p: &mut &T                                                         |  

 real, m ⊢ p: &mut T                                                              | inh, m ⊢ p: &mut T
 ------------------------------ "Deref"                                           | ------------------------------ "Deref(EatInner)"
 inh, m ⊢ &mut p: &mut &mut T                                                     | inh, m ⊢ &mut p: &mut &mut T

                                                                                  | inh, m ⊢ p: &T
                                                                                  | -------------------------- "Deref(EatInner)"
                                                                                  | inh, m ⊢ &mut p: &&mut T

 real, ro ⊢ p: T                                                                  | real, ro ⊢ p: T
 ---------------------- "DerefMutWithShared"                                      | ---------------------- "DerefMutWithShared(EatOuter)"
 real, m ⊢ &p: &mut T                                                             | real, m ⊢ &p: &mut T

 _, ro ⊢ p: T                                                                     | _, ro ⊢ p: T
 ------------------------------------------- "DerefMutWithShared"                 | ------------------------------------------- "DerefMutWithShared(EatOuter)"
 inh, m ⊢ &p: &mut T, T is not a reference                                        | inh, m ⊢ &p: &mut T, T is not a reference

 real, ro ⊢ p: &T                                                                 | inh, ro ⊢ p: &mut T
 ---------------------- "DerefMutWithShared"                                      | ---------------------- "Deref(EatInner)"
 inh, m ⊢ &p: &mut &T                                                             | inh, m ⊢ &p: &mut &T

 real, ro ⊢ p: &mut T                                                             | inh, ro ⊢ p: &mut T
 -------------------------- "DerefMutWithShared"                                  | -------------------------- "DerefMutWithShared(EatInner)"
 inh, m ⊢ &p: &mut &mut T                                                         | inh, m ⊢ &p: &mut &mut T

 inh, m ⊢ x: &T                                                                   | inh, m ⊢ x: &T
 -------------------- "BindingBorrow"                                             | -------------------- "BindingBorrow"
 real, m ⊢ ref x: T                                                               | real, m ⊢ ref x: T

 inh, m ⊢ x: &mut T                                                               | inh, m ⊢ x: &mut T
 ------------------------ "BindingBorrow"                                         | ------------------------ "BindingBorrow"
 real, m ⊢ ref mut x: T                                                           | real, m ⊢ ref mut x: T

                                                                                  | 
 ------------- "Binding"                                                          | ------------- "Binding"
 r, m ⊢ x: T                                                                      | r, m ⊢ x: T

                                                                                  | 
 ----------------- "Binding"                                                      | ----------------- "Binding"
 r, m ⊢ mut x: T                                                                  | r, m ⊢ mut x: T


inherited_ref_on_ref: "EatOuter" -> "EatInner"
